android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/ScopedFirebaseCoreService.java

```diff
@@ -0,0 +1,163 @@
+package versioned.host.exp.exponent.modules.universal;
+
+import android.content.Context;
+import android.util.Base64;
+
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import host.exp.exponent.kernel.ExperienceId;
+
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.FirebaseOptions;
+
+import org.unimodules.core.ModuleRegistry;
+
+import expo.modules.firebase.core.FirebaseCoreService;
+import expo.modules.firebase.core.FirebaseCoreOptions;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.List;
+
+import java.io.UnsupportedEncodingException;
+
+public class ScopedFirebaseCoreService extends FirebaseCoreService {
+
+  private ModuleRegistry mModuleRegistry;
+  private String mProtectedAppName;
+  private String mAppName;
+  private FirebaseOptions mAppOptions;
+
+  public ScopedFirebaseCoreService(Context context, JSONObject manifest, ExperienceId experienceId) {
+    super(context);
+
+    // Get the default firebase app name
+    FirebaseApp defaultApp = getFirebaseApp(null);
+    mProtectedAppName = (defaultApp != null) ? defaultApp.getName() : "[DEFAULT]";
+
+    // Get experience key & unique app name
+    String experienceKey = getEncodedExperienceId(experienceId);
+    mAppName = "__sandbox_" + experienceKey;
+    mAppOptions = getOptionsFromManifest(manifest);
+
+    // Delete all previously created apps, except for the "default" one
+    // which will be updated/created/deleted only when it has changed
+    List<FirebaseApp> apps = FirebaseApp.getApps(context);
+    for (FirebaseApp app : apps) {
+      if (!app.getName().equals(mProtectedAppName) && !app.getName().equals(mAppName)) {
+        app.delete();
+      }
+    }
+
+    // Initialize the firebase app. This will delete/create/update the app
+    // if it has changed, and leaves the app untouched when the config
+    // is the same.
+    updateFirebaseApp(mAppOptions, mAppName);
+  }
+
+  private static String getEncodedExperienceId(ExperienceId experienceId) {
+    try {
+      String encodedUrl = experienceId.getUrlEncoded();
+      byte[] data = encodedUrl.getBytes("UTF-8");
+      String base64 = Base64.encodeToString(data, Base64.URL_SAFE | Base64.NO_PADDING);
+      return base64.replace("\n", "").replace("\r", "");
```

Comment:
[`Base64.NO_WRAP`](https://developer.android.com/reference/android/util/Base64.html#NO_WRAP) should take care of that

Reply:
I tried that, but I still got a new-line at the end of the string I believe. 🤔

Reply:
This could be mitigated using `String#trim()`, which somehow seems more stable than replacing these characters manually, but that may be just me

Reply:
Tried `Base64.NO_WRAP` and it indeed seems to work as expected and as documented. Must have been a fluke when I tested that earlier. Thanks for the suggestion 👍 👍